# Recurrent Neural Networks and Language Models

In [5]:
import torch
print(torch.__version__)


2.2.0+cu121


In [6]:
import torch
import torch.nn as nn
import torch.optim as optim
import math
from tqdm import tqdm
import pickle

In [7]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(device)

cuda


In [8]:
# seed for reproducability
SEED = 122
torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

# Load Data
Source Credit: https://www.kaggle.com/datasets/prashantkarwasra/books-dataset-text-generation?select=LOTR1.txt

The dataset is taken from kaggle. To clean the data we removed all tab spaces, and remove all special characters except , . ! ?. 

We have out final output as a single line text file with the book contents.



In [9]:
from datasets import Dataset

harry_potter_dataset = "./data/LOTR1_clean.txt"

# Read the data from the file
with open(harry_potter_dataset, 'r') as f:
    data = f.read()

# Creating list of dictionaries
data = data.split(".")
data = [{"text": row} for row in data]

# Creating dataset object
dataset = Dataset.from_list(data)
dataset

Dataset({
    features: ['text'],
    num_rows: 11149
})

In [10]:
from datasets import DatasetDict

train_test = dataset.train_test_split(test_size=0.2)

# 10% test set and 10% validation set
train_test_valid = train_test['test'].train_test_split(test_size=0.5)

dataset = DatasetDict({
    'train': train_test['train'],
    'test': train_test_valid['test'],
    'validation': train_test_valid['train']})

dataset

DatasetDict({
    train: Dataset({
        features: ['text'],
        num_rows: 8919
    })
    test: Dataset({
        features: ['text'],
        num_rows: 1115
    })
    validation: Dataset({
        features: ['text'],
        num_rows: 1115
    })
})

In [7]:
print(dataset['train'][333]['text']) # empty string

'''
If you try to change the index you might notice that sometimes there is no paragraph 
and rather an empty string so we will have to care of that later.
'''

 The air seemed heavy and the making of words wearisome


'\nIf you try to change the index you might notice that sometimes there is no paragraph \nand rather an empty string so we will have to care of that later.\n'

# Preprocessing

## Tokenizing

In [11]:
from torchtext.data.utils import get_tokenizer

In [12]:
tokenizer = get_tokenizer('basic_english')

#function to tokenize
tokenize_data = lambda example, tokenizer: {'tokens': tokenizer(example['text'])}  

#map the function to each example
tokenized_dataset = dataset.map(tokenize_data, remove_columns=['text'], fn_kwargs={'tokenizer': tokenizer})
print(tokenized_dataset['train'][33]['tokens'])

Map:   0%|          | 0/8919 [00:00<?, ? examples/s]

Map:   0%|          | 0/1115 [00:00<?, ? examples/s]

Map:   0%|          | 0/1115 [00:00<?, ? examples/s]

['at', 'last', 'they', 'got', 'the', 'idea', 'that', 'enemies', 'were', 'in', 'buckland', ',', 'some', 'strange', 'invasion', 'from', 'the', 'old', 'forest']


## Numericializing

In [13]:
from torchtext.vocab import build_vocab_from_iterator

In [14]:
## numericalizing
vocab = build_vocab_from_iterator(tokenized_dataset['train']['tokens'], 
min_freq=3) 
vocab.insert_token('<unk>', 0)           
vocab.insert_token('<eos>', 1)            
vocab.set_default_index(vocab['<unk>'])   
print(len(vocab))                         
print(vocab.get_itos()[:10])    

3429
['<unk>', '<eos>', ',', 'the', 'and', 'of', 'to', 'a', 'he', 'in']


In [16]:
with open('model/vocab_lm.pkl', 'wb') as f:
    pickle.dump(vocab, f)

After loading and splitting the personally chosen dataset a DatasetDictonary is created. Then on that object the preprocessing steps are applied. Firstly, we tokenize the dataset using torchtext's `get_tokenizer`. The `tokenize_data` function is applied to each example where the `text` column is removed and a new `tokens` column containing the tokenized text is added.

Then our vocabulary is made using the `build_vocab_from_iterator` method from torchtext. We use the training dataset, and consider words that has occured at least three times. This is done to make sure that our vocab does not get too big. Then we add `<unk>` to signify unknown and `<eos>` to signify end of sentence. After all this the vocab size came out to be 3429.

## Prepare the batch loader

### Prepare data

In [19]:
def get_data(dataset, vocab, batch_size):
    data = []                                                   
    for example in dataset:
        if example['tokens']:         
            #appends eos so we know it ends....so model learn how to end...                             
            tokens = example['tokens'].append('<eos>')   
            #numericalize          
            tokens = [vocab[token] for token in example['tokens']] 
            data.extend(tokens)                                    
    data = torch.LongTensor(data)                                 
    num_batches = data.shape[0] // batch_size #get the int number of batches...
    data = data[:num_batches * batch_size] #make the batch evenly, and cut out any remaining                      
    data = data.view(batch_size, num_batches)          
    return data #[batch size, bunch of tokens]

In [20]:
batch_size = 128
train_data = get_data(tokenized_dataset['train'], vocab, batch_size)
valid_data = get_data(tokenized_dataset['validation'], vocab, batch_size)
test_data  = get_data(tokenized_dataset['test'], vocab, batch_size)

## Modeling

In [21]:
class LSTMLanguageModel(nn.Module):
    def __init__(self, vocab_size, emb_dim, hid_dim, num_layers, dropout_rate):
                
        super().__init__()
        self.num_layers = num_layers
        self.hid_dim = hid_dim
        self.emb_dim = emb_dim

        self.embedding = nn.Embedding(vocab_size, emb_dim)
        self.lstm = nn.LSTM(emb_dim, hid_dim, num_layers=num_layers, 
                    dropout=dropout_rate, batch_first=True)
        self.dropout = nn.Dropout(dropout_rate)
        self.fc = nn.Linear(hid_dim, vocab_size)
        
        self.init_weights()
        
    def init_weights(self):
        init_range_emb = 0.1
        init_range_other = 1/math.sqrt(self.hid_dim)
        self.embedding.weight.data.uniform_(-init_range_emb, init_range_emb)
        self.fc.weight.data.uniform_(-init_range_other, init_range_other)
        self.fc.bias.data.zero_()
        for i in range(self.num_layers):
            self.lstm.all_weights[i][0] = torch.FloatTensor(self.emb_dim,
                    self.hid_dim).uniform_(-init_range_other, init_range_other) 
            self.lstm.all_weights[i][1] = torch.FloatTensor(self.hid_dim, 
                    self.hid_dim).uniform_(-init_range_other, init_range_other) 

    def init_hidden(self, batch_size, device):
        hidden = torch.zeros(self.num_layers, batch_size, self.hid_dim).to(device)
        cell   = torch.zeros(self.num_layers, batch_size, self.hid_dim).to(device)
        return hidden, cell
    
    def detach_hidden(self, hidden):
        hidden, cell = hidden
        hidden = hidden.detach()
        cell = cell.detach()
        return hidden, cell

    def forward(self, src, hidden):
        #src: [batch size, seq len]
        embedding = self.dropout(self.embedding(src))
        #embedding: [batch size, seq len, emb_dim]
        output, hidden = self.lstm(embedding, hidden)      
        #output: [batch size, seq len, hid_dim]
        #hidden = h, c = [num_layers * direction, seq len, hid_dim)
        output = self.dropout(output) 
        prediction = self.fc(output)
        #prediction: [batch size, seq_len, vocab size]
        return prediction, hidden

The LSTM contains memory cells and gates that allows it to selectively remember and forget information from previous times steps. The input gate controls the flow of information from the previous memeory cell to the current memory cell. The forget gate controls the flow of information from previous cell to the current cell and allows it to selectively forget or remember information from previous steps. The memory cell stores information that can be selectively modified by the input and forget gates. And the output gate controls the flow of information from the memory cell to the hidden state and output.

Our defined class is firstly converting tokens into embeddings then processing those embeddings with stacked LSTM layers to capture temporal dependencies (relationships between past and future events or states in a time series). Then we apply dropout to embeddings to prevent overfitting by randomly zero-ing inputs during training. Then finally we use a linear layer to map LSTM outputs to vocabulary logits for predicting the next word.

## Training

Follows very basic procedure.  One note is that some of the sequences that will be fed to the model may involve parts from different sequences in the original dataset or be a subset of one (depending on the decoding length). For this reason we will reset the hidden state every epoch, this is like assuming that the next batch of sequences is probably always a follow up on the previous in the original dataset.

In [22]:
vocab_size = len(vocab)
emb_dim = 1024                # 400 in the paper
hid_dim = 1024                # 1150 in the paper
num_layers = 2                # 3 in the paper
dropout_rate = 0.65              
lr = 1e-3                     

In [23]:
model = LSTMLanguageModel(vocab_size, emb_dim, hid_dim, num_layers, dropout_rate).to(device)
optimizer = optim.Adam(model.parameters(), lr=lr)
criterion = nn.CrossEntropyLoss()
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print(f'The model has {num_params:,} trainable parameters')

The model has 23,819,621 trainable parameters


In [24]:
def get_batch(data, seq_len, idx):
    #data #[batch size, bunch of tokens]
    src    = data[:, idx:idx+seq_len]                   
    target = data[:, idx+1:idx+seq_len+1]  #target simply is ahead of src by 1            
    return src, target

In [25]:
def train(model, data, optimizer, criterion, batch_size, seq_len, clip, device):
    
    epoch_loss = 0
    model.train()
    # drop all batches that are not a multiple of seq_len
    # data #[batch size, bunch of tokens]
    num_batches = data.shape[-1]
    data = data[:, :num_batches - (num_batches -1) % seq_len]  #we need to -1 because we start at 0
    num_batches = data.shape[-1]
    
    #reset the hidden every epoch
    hidden = model.init_hidden(batch_size, device)
    
    for idx in tqdm(range(0, num_batches - 1, seq_len), desc='Training: ',leave=False):
        optimizer.zero_grad()
        
        #hidden does not need to be in the computational graph for efficiency
        hidden = model.detach_hidden(hidden)

        src, target = get_batch(data, seq_len, idx) #src, target: [batch size, seq len]
        src, target = src.to(device), target.to(device)
        batch_size = src.shape[0]
        prediction, hidden = model(src, hidden)               

        #need to reshape because criterion expects pred to be 2d and target to be 1d
        prediction = prediction.reshape(batch_size * seq_len, -1)  #prediction: [batch size * seq len, vocab size]  
        target = target.reshape(-1)
        loss = criterion(prediction, target)
        
        loss.backward()
        torch.nn.utils.clip_grad_norm_(model.parameters(), clip)
        optimizer.step()
        epoch_loss += loss.item() * seq_len
    return epoch_loss / num_batches

In [26]:
def evaluate(model, data, criterion, batch_size, seq_len, device):

    epoch_loss = 0
    model.eval()
    num_batches = data.shape[-1]
    data = data[:, :num_batches - (num_batches -1) % seq_len]
    num_batches = data.shape[-1]

    hidden = model.init_hidden(batch_size, device)

    with torch.no_grad():
        for idx in range(0, num_batches - 1, seq_len):
            hidden = model.detach_hidden(hidden)
            src, target = get_batch(data, seq_len, idx)
            src, target = src.to(device), target.to(device)
            batch_size= src.shape[0]

            prediction, hidden = model(src, hidden)
            prediction = prediction.reshape(batch_size * seq_len, -1)
            target = target.reshape(-1)

            loss = criterion(prediction, target)
            epoch_loss += loss.item() * seq_len
    return epoch_loss / num_batches

In [27]:
n_epochs = 50
seq_len  = 50 #<----decoding length
clip    = 0.25

lr_scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, factor=0.5, patience=0)

best_valid_loss = float('inf')

for epoch in range(n_epochs):
    train_loss = train(model, train_data, optimizer, criterion, 
                batch_size, seq_len, clip, device)
    valid_loss = evaluate(model, valid_data, criterion, batch_size, 
                seq_len, device)

    lr_scheduler.step(valid_loss)

    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        torch.save(model.state_dict(), 'model/best-val-lstm_lm.pt')

    print(f'\tTrain Perplexity: {math.exp(train_loss):.3f}')
    print(f'\tValid Perplexity: {math.exp(valid_loss):.3f}')

	Train Perplexity: 541.965
	Valid Perplexity: 312.784


	Train Perplexity: 358.620
	Valid Perplexity: 282.367


	Train Perplexity: 301.733
	Valid Perplexity: 232.250


	Train Perplexity: 246.375
	Valid Perplexity: 190.738


	Train Perplexity: 203.085
	Valid Perplexity: 164.923


	Train Perplexity: 178.122
	Valid Perplexity: 149.764


	Train Perplexity: 160.107
	Valid Perplexity: 138.014


	Train Perplexity: 145.937
	Valid Perplexity: 128.694


	Train Perplexity: 135.543
	Valid Perplexity: 122.259


	Train Perplexity: 127.478
	Valid Perplexity: 116.618


	Train Perplexity: 120.509
	Valid Perplexity: 113.102


	Train Perplexity: 114.182
	Valid Perplexity: 108.680


	Train Perplexity: 108.599
	Valid Perplexity: 105.399


	Train Perplexity: 103.807
	Valid Perplexity: 102.296


	Train Perplexity: 99.715
	Valid Perplexity: 99.890


	Train Perplexity: 95.758
	Valid Perplexity: 98.295


	Train Perplexity: 91.913
	Valid Perplexity: 96.574


	Train Perplexity: 88.592
	Valid Perplexity: 95.199


	Train Perplexity: 85.634
	Valid Perplexity: 93.513


	Train Perplexity: 82.942
	Valid Perplexity: 92.301


	Train Perplexity: 80.857
	Valid Perplexity: 91.766


	Train Perplexity: 78.155
	Valid Perplexity: 91.038


	Train Perplexity: 76.005
	Valid Perplexity: 90.127


	Train Perplexity: 74.024
	Valid Perplexity: 89.863


	Train Perplexity: 71.466
	Valid Perplexity: 89.070


	Train Perplexity: 69.903
	Valid Perplexity: 89.400


	Train Perplexity: 68.245
	Valid Perplexity: 91.574


	Train Perplexity: 67.240
	Valid Perplexity: 90.093


	Train Perplexity: 66.278
	Valid Perplexity: 90.069


	Train Perplexity: 65.002
	Valid Perplexity: 90.103


	Train Perplexity: 65.580
	Valid Perplexity: 90.014


	Train Perplexity: 64.589
	Valid Perplexity: 90.011


	Train Perplexity: 64.738
	Valid Perplexity: 90.007


	Train Perplexity: 65.492
	Valid Perplexity: 90.008


	Train Perplexity: 64.687
	Valid Perplexity: 90.005


	Train Perplexity: 64.462
	Valid Perplexity: 90.001


	Train Perplexity: 65.363
	Valid Perplexity: 90.000


	Train Perplexity: 64.592
	Valid Perplexity: 89.999


	Train Perplexity: 64.656
	Valid Perplexity: 89.999


	Train Perplexity: 65.258
	Valid Perplexity: 89.999


	Train Perplexity: 64.528
	Valid Perplexity: 89.999


	Train Perplexity: 64.448
	Valid Perplexity: 89.999


	Train Perplexity: 64.689
	Valid Perplexity: 89.999


	Train Perplexity: 64.635
	Valid Perplexity: 89.999


	Train Perplexity: 65.396
	Valid Perplexity: 89.999


	Train Perplexity: 64.613
	Valid Perplexity: 89.999


	Train Perplexity: 65.494
	Valid Perplexity: 89.999


	Train Perplexity: 64.519
	Valid Perplexity: 89.999


	Train Perplexity: 65.592
	Valid Perplexity: 89.999


	Train Perplexity: 64.643
	Valid Perplexity: 89.999


Firstly, we initialize hyperparameters such as vocab size, embedding dimensions, hidden dimensions, number of layers, dropout rate and learning rate. We then move the model to device of user's choice (GPU in our case). Then adam optimizer is defined to optimize the model's parameters and CrossEntropyLoss criterion is use to compute the loss during training.

We train for 50 epochs. In each epoch, the training data is deivided into batches of specific sequece using the `get_batch` method that we defined. At the start of each epoch hidden state is reset and for each batch the model parameters are zeroed and forward pass is done. The loss is then calculated by using the predicted probabilites for a token compared with actual next token. The gradient is calculated using backpropagation and the model's parameters are updated using the optimizer. We also keep track of loss for each epoch.

At the end of an epoch, the model is put into eval mode and the validation data is processed with same process as stated before and validation loss is calculated. The learning rate scheduler is used to adjust the learning rate based on the validation loss. Finally, the model params are saved if the current validation loss is the best one observed so far. 

## Testing

In [28]:
# Load the best model state from the saved checkpoint
model.load_state_dict(torch.load('model/best-val-lstm_lm.pt',  map_location=device))

# Evaluate the model on the test data
test_loss = evaluate(model, test_data, criterion, batch_size, seq_len, device)

# Print the test perplexity
print(f'Test Perplexity: {math.exp(test_loss):.3f}')

Test Perplexity: 90.490


# Real-world inference

In [29]:
def generate(prompt, max_seq_len, temperature, model, tokenizer, vocab, device, seed=None):
    if seed is not None:
        torch.manual_seed(seed)
    model.eval()
    tokens = tokenizer(prompt)
    indices = [vocab[t] for t in tokens]
    batch_size = 1
    hidden = model.init_hidden(batch_size, device)
    with torch.no_grad():
        for i in range(max_seq_len):
            src = torch.LongTensor([indices]).to(device)
            prediction, hidden = model(src, hidden)
            
            #prediction: [batch size, seq len, vocab size]
            #prediction[:, -1]: [batch size, vocab size] #probability of last vocab
            
            probs = torch.softmax(prediction[:, -1] / temperature, dim=-1)  
            prediction = torch.multinomial(probs, num_samples=1).item()    
            
            while prediction == vocab['<unk>']: #if it is unk, we sample again
                prediction = torch.multinomial(probs, num_samples=1).item()

            if prediction == vocab['<eos>']:    #if it is eos, we stop
                break

            indices.append(prediction) #autoregressive, thus output becomes input

    itos = vocab.get_itos()
    tokens = [itos[i] for i in indices]
    return tokens

In [34]:
prompt = 'Bilbo Baggins is'
max_seq_len = 30
seed = 0

#smaller the temperature, more diverse tokens but comes 
#with a tradeoff of less-make-sense sentence
temperatures = [0.5, 0.7, 0.75, 0.8, 1.0]
for temperature in temperatures:
    generation = generate(prompt, max_seq_len, temperature, model, tokenizer, 
                          vocab, device, seed)
    print(str(temperature)+'\n'+' '.join(generation)+'\n')

0.5
bilbo baggins is a hobbit

0.7
bilbo baggins is a hole that he had not had than a others , and he was the old morning

0.75
bilbo baggins is sitting down near the gate , and it could not be able to find you , and i will do you , said gandalf

0.8
bilbo baggins is sitting down near the gate , and it could not be able to find you , and i will do you , said gandalf

1.0
bilbo baggins is sitting down near the hand



# Web application interface documentation

I made the interface using Dash. The entire UI along with necessary model implemenetations are carried out in `app.py` file. It is a simple UI with a query field, search button, some validations and a result section. The demo can be found in the `README.md` file of A2 folder.

The model is integrated with the interface with simple steps. Firstly, the vocab is loaded using `pickle` and the model is setup using pre-trained weights into it. Temperature settings are passed to the language model during text generation to control the diversity. The model completes a sentence based on user's prompts.

The user interaction flow is as follows:
- User enters a prompt into the field
- Selects the Tempratures 
- Adjusts the Token length
- Clicks the button
- Texts are generated completing the prompt and are showed in proper manner

Also, since the model is trained only on lord of the rings: fellowship of the ring - if the user enters anything other than the context relating to lord of the rings, the results wont make much sense.